demucs/stft.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (C) 2025 Mixxx Development Team
+# All rights reserved.
+#
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+# First author is Anmol Mishra.
+"""
+This module implements a custom STFT process that is compatible with ONNX export.
+It uses PyTorch's convolution operations to compute the STFT, avoiding the use of complex
+numbers directly, which can be problematic for ONNX export.
+"""
+import torch
+
+# Constants set for Demucs
+NFFT = 4096                                         # Number of FFT components for the STFT process
+HOP_LENGTH = 1024                                   # Number of samples between successive frames in the STFT
+WINDOW_TYPE = 'hann'                                # Type of window function used in the STFT
+WINDOW_LENGTH = NFFT                                # Length of the window function
+NORMALIZED = True                                   # Whether to normalize the window function
+CENTER = True                                       # Whether to center the input signal before STFT
+PAD_MODE = 'reflect'                                # Select reflect or constant padding mode for the STFT process.
+
+# Create window function lookup once
+WINDOW_FUNCTIONS = {
+    'bartlett': torch.bartlett_window,
+    'blackman': torch.blackman_window,
+    'hamming': torch.hamming_window,
+    'hann': torch.hann_window,
+    'kaiser': lambda x: torch.kaiser_window(x, periodic=True, beta=12.0)
+}
+# Define default window function
+DEFAULT_WINDOW_FN = torch.hann_window
+
+class STFT_Process(torch.nn.Module):
+    def __init__(self, n_fft=NFFT, hop_len=HOP_LENGTH, window_type=WINDOW_TYPE, window_length=WINDOW_LENGTH, normalized=NORMALIZED, center=CENTER, pad_mode=PAD_MODE):
+        super(STFT_Process, self).__init__()
+        self.n_fft = n_fft
+        self.hop_len = hop_len
+        self.window_type = window_type
+        self.window_length = window_length
+        self.normalized = normalized
+        self.center = center
+        self.pad_mode = pad_mode
+        self.half_n_fft = n_fft // 2  # Precompute once
+        
+        # Get window function and compute window once
+        window = WINDOW_FUNCTIONS.get(window_type, DEFAULT_WINDOW_FN)(self.window_length).float()
+        if self.window_length != self.n_fft:
+            raise NotImplementedError(f"Window length {self.window_length} does not match n_fft {self.n_fft}.")
+
+        # STFT forward pass preparation
+        time_steps = torch.arange(self.n_fft, dtype=torch.float32).unsqueeze(0)
+        frequencies = torch.arange(self.half_n_fft + 1, dtype=torch.float32).unsqueeze(1)
+        
+        # Calculate omega matrix once
+        omega = 2 * torch.pi * frequencies * time_steps / n_fft
+
+        # Calculate window function
+        cos_kernel = (torch.cos(omega) * window.unsqueeze(0)).unsqueeze(1)
+        sin_kernel = (-torch.sin(omega) * window.unsqueeze(0)).unsqueeze(1)
```

Comment:
Could you measure if this runs faster:
```suggestion
        window_unsq = window.unsqueeze(0)
        cos_kernel = (torch.cos(omega) * window_unsq).unsqueeze(1)
        sin_kernel = (-torch.sin(omega) * window_unsq).unsqueeze(1)
```

Reply:
They're almost the same time, I measured with `torch.autograd.profiler.profile`:
```
STFT Process with window_unsq:
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  
           aten::sin        36.65%       9.953ms        36.65%       9.953ms       9.953ms             1  
           aten::cos        28.84%       7.830ms        28.84%       7.830ms       7.830ms             1  
           aten::mul        19.10%       5.187ms        19.10%       5.187ms       2.594ms             2  
           aten::neg        15.30%       4.153ms        15.30%       4.153ms       4.153ms             1  
     aten::unsqueeze         0.09%      23.209us         0.11%      29.542us       9.847us             3  
    aten::as_strided         0.02%       6.333us         0.02%       6.333us       2.111us             3  
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 27.152ms

STFT Process with window.unsqueeze(0):
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  
           aten::sin        36.95%       8.451ms        36.95%       8.451ms       8.451ms             1  
           aten::cos        36.93%       8.448ms        36.93%       8.448ms       8.448ms             1  
           aten::mul        21.94%       5.018ms        21.94%       5.018ms       2.509ms             2  
           aten::neg         4.09%     934.594us         4.09%     934.594us     934.594us             1  
     aten::unsqueeze         0.07%      16.623us         0.10%      22.708us       5.677us             4  
    aten::as_strided         0.03%       6.085us         0.03%       6.085us       1.521us             4  
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 22.874ms
```

Reply:
Isn't the `Self CPU time total` what matters here? Than the second variant would be clearly better.

Reply:
The impact due to `unsqueeze` operations seems minimal, but I observed similar trend with `CPU Time total` on running this test a few times. So we can keep the second variant.

Another aspect is this code is inside constructor so it will not be used during forward pass and won't be in the exported ONNX graph. But it does affect model load times when called inside python.